android/app/src/main/java/com/mattermost/helpers/CustomPushNotificationHelper.java

```diff
@@ -227,6 +238,117 @@ public static void createNotificationChannels(Context context) {
+            return false;
+        }
+
+        WMDatabase db = getDatabaseForServer(dbHelper, context, serverUrl);
+        if (db == null) {
+            Log.i("Mattermost Notifications Signature verification", "Cannot access the server database");
+            return false;
+        }
+
+        if (signature == "NO_SIGNATURE") {
+            String version = queryConfigServerVersion(db);
+            if (version == null) {
+                Log.i("Mattermost Notifications Signature verification", "No server version");
+                return false;
+            }
+
+            if (!version.matches("[0-9]+(\\.[0-9]+)*")) {
+                Log.i("Mattermost Notifications Signature verification", "Invalid server version");
+                return false;
+            }
+
+            String[] parts = version.split("\\.");
+            int mayor = parts.length > 0 ? Integer.parseInt(parts[0]) : 0;
+            int minor = parts.length > 1 ? Integer.parseInt(parts[1]) : 0;
+            int patch = parts.length > 2 ? Integer.parseInt(parts[2]) : 0;
+            int mayorTarget = 9;
+            int minorTarget = 7;
+            int patchTarget = 0;
+
+            if (mayor < mayorTarget) {
+                return true;
+            }
+            if (mayor == mayorTarget && minor < minorTarget) {
+                return true;
+            }
+            if (mayor == mayorTarget && minor == minorTarget && patch < patchTarget) {
+                return true;
+            }
+
+            Log.i("Mattermost Notifications Signature verification", "Server version should send signature");
+            return false;
+        }
+
+        String signingKey = queryConfigSigningKey(db);
+        if (signingKey == null) {
+            Log.i("Mattermost Notifications Signature verification", "No signing key");
+            return false;
+        }
+
+        try {
+            byte[] encoded = Base64.decode(signingKey, 0);
+            KeyFactory kf = KeyFactory.getInstance("EC");
+            PublicKey pubKey = (PublicKey) kf.generatePublic(new X509EncodedKeySpec(encoded));
+
+            String storedDeviceToken = getDeviceToken(dbHelper);
+            if (storedDeviceToken == null) {
+                Log.i("Mattermost Notifications Signature verification", "No device token stored");
+                return false;
+            }
+            String[] tokenParts = storedDeviceToken.split(":", 2);
+            if (tokenParts.length != 2) {
+                Log.i("Mattermost Notifications Signature verification", "Wrong stored device token format");
+                return false;
+            }
+            String deviceToken = tokenParts[1].substring(0, tokenParts[1].length() -1 );
+            if (deviceToken.isEmpty()) {
+                Log.i("Mattermost Notifications Signature verification", "Empty stored device token");
+                return false;
+            }
+
+            Jwts.parser()
+                    .require("ack_id", ackId)
```

Comment:
Nit: do we actually need to verify the ack_id value? No harm in it doing it, it just seems unnecessary to me.

Reply:
Discussed in the channel